File: servers/go/internal/domain/member/port.go
package member

import "context"

type ReadRepository interface {
	GetByID(ctx context.Context, id MemberID) (*Member, error)
	GetByIdent(ctx context.Context, ident MemberIdent) (*Member, error)
}

type WriteRepository interface {
	Create(ctx context.Context, prepared MemberPrepared) (*Member, error)
	GetByIdentForUpdate(ctx context.Context, ident MemberIdent) (*Member, error)
	UpdateStatus(ctx context.Context, id MemberID, status MemberStatus) error
}

File: server/crates/domain/src/member/port.rs
use async_trait::async_trait;

use crate::PortResult;

use super::{Member, MemberId, MemberIdent, MemberPrepared, MemberStatus};

#[async_trait]
pub trait MemberWriteRepoPort: Send {
    async fn create(&mut self, insert: &MemberPrepared) -> PortResult<Member>;
    async fn get_by_ident_for_update(&mut self, ident: &MemberIdent) -> PortResult<Option<Member>>;
    async fn update_status(&mut self, id: MemberId, status: MemberStatus) -> PortResult<()>;
}

#[async_trait]
pub trait MemberReadRepoPort: Send + Sync {
    async fn get_by_id(&self, id: MemberId) -> PortResult<Option<Member>>;
    async fn get_by_ident(&self, ident: &MemberIdent) -> PortResult<Option<Member>>;
}
